sdk/js/src/token_bridge/redeem.ts

```diff
@@ -219,3 +235,121 @@ export async function redeemOnAlgorand(
+export async function redeemOnNear(
+  client: nearAccount,
+  tokenBridge: string,
+  vaa: Uint8Array
+): Promise<String> {
+  let p = _parseVAAAlgorand(vaa);
+
+  if (p.ToChain !== CHAIN_ID_NEAR) {
+    throw new Error("Not destined for NEAR");
+  }
+
+  let user = await client.viewFunction(tokenBridge, "hash_lookup", {
+    hash: uint8ArrayToHex(p.ToAddress as Uint8Array),
+  });
+
+  if (!user[0]) {
+    throw new Error(
+      "Unregistered receiver (receiving account is not registered)"
+    );
+  }
+
+  user = user[1];
+
+  let token = await getForeignAssetNear(
+    client,
+    tokenBridge,
+    p.FromChain as ChainId,
+    p.Contract as string
+  );
+
+  if (token === "") {
+    throw new Error("Unregistered token (this been attested yet?)");
+  }
+
+  if (
+    (p.Contract as string) !==
+    "0000000000000000000000000000000000000000000000000000000000000000"
+  ) {
+    let bal = await client.viewFunction(token as string, "storage_balance_of", {
+      account_id: user,
+    });
+
+    if (bal === null) {
+      console.log("Registering ", user, " for ", token);
+      bal = nearProviders.getTransactionLastResult(
+        await client.functionCall({
+          contractId: token as string,
+          methodName: "storage_deposit",
+          args: { account_id: user, registration_only: true },
+          gas: new BN("100000000000000"),
+          attachedDeposit: new BN("2000000000000000000000"), // 0.002 NEAR
+        })
+      );
+    }
+
+    if (
+      p.Fee !== undefined &&
+      Buffer.compare(
+        p.Fee,
+        Buffer.from(
+          "0000000000000000000000000000000000000000000000000000000000000000",
+          "hex"
+        )
+      ) !== 0
+    ) {
+      let bal = await client.viewFunction(
+        token as string,
+        "storage_balance_of",
+        {
+          account_id: client.accountId,
+        }
+      );
+
+      if (bal === null) {
+        console.log("Registering ", client.accountId, " for ", token);
+        bal = nearProviders.getTransactionLastResult(
+          await client.functionCall({
+            contractId: token as string,
+            methodName: "storage_deposit",
+            args: { account_id: client.accountId, registration_only: true },
+            gas: new BN("100000000000000"),
+            attachedDeposit: new BN("2000000000000000000000"), // 0.002 NEAR
+          })
+        );
+      }
+    }
+  }
+
+  let result = await client.functionCall({
+    contractId: tokenBridge,
+    methodName: "submit_vaa",
+    args: {
+      vaa: uint8ArrayToHex(vaa),
+    },
+    attachedDeposit: new BN("100000000000000000000000"),
+    gas: new BN("150000000000000"),
+  });
+
+  result = await client.functionCall({
+    contractId: tokenBridge,
+    methodName: "submit_vaa",
+    args: {
+      vaa: uint8ArrayToHex(vaa),
+    },
+    attachedDeposit: new BN("100000000000000000000000"),
+    gas: new BN("150000000000000"),
+  });
+
+  return nearProviders.getTransactionLastResult(result);
+}
```

Comment:
You are doing way more work in the redeem than any of the redeems for the other chains.  Is this the new standard?  Should the others follow this model?  (I'm not asking you to change the others.) I'm looking to make Injective consistent with the proper way of doing things.

Reply:
failures on near are unbelievably hard to handle without opening yourself to exploits.  So,  I work super hard to make it super hard to fail